Major Refactor: Version 1.0.0 with Config Flow and UI integration

alnavasa opened the pull request:
**Major Refactor: Version 0.2.0 (Breaking Changes, Config Flow, Cameras, Firmware, Diagnostics)**

## 🚀 Major Refactor - Version 0.2.0

### Highlights
- 🔧 Migrated to config flow (no more YAML configuration)
- 📷 Camera support (live and snapshot)
- 🔍 Firmware update support
- 📈 Over 50 sensors including diagnostics, duration, and network status
- 🌍 Translations added: EN, ES, DE, FR, IT, NL
- 🧩 Compatible with modern Home Assistant standards

### ⚠️ Breaking Changes
- YAML configuration is no longer supported.
- Users must remove the old configuration and set up the integration via the UI (Config Flow).

### Screenshots & Lovelace Example
![setup](https://github.com/user-attachments/assets/f9474951-5746-493b-9c60-3cb48b4142a6)

<img width="442" alt="lovelace" src="https://github.com/user-attachments/assets/69803d9d-9d6b-40fc-98e3-98789b872fe1" />
<img width="1063" alt="sensors" src="https://github.com/user-attachments/assets/33340619-8654-40a4-992e-54f9b1c13ae5" />


### Credits
- Original by [@jellespijker](https://github.com/jellespijker)
- Enhanced by [@alnavasa](https://github.com/alnavasa)

Feedback and testing welcome!

alnavasa committed:
Major Refactor - Version 0.2.0 (Breaking Changes)

This PR introduces a complete refactor of the Ultimaker integration:

### 🔧 Changes
- UI-based setup with `config_flow`
- Full rewrite using `DataUpdateCoordinator`
- New sensor structure with 60+ entities
- Camera support (stream + snapshot)
- Update entity for firmware
- Entity registry settings improved
- Device grouping, icon updates, translations

### ⚠️ Breaking Changes
- YAML configuration removed
- Old entities may be renamed or removed
- Users must remove the previous integration and re-add it via the UI

Tested on Ultimaker S5 with firmware 9.x.
Co-authored with @alnavasa.

alnavasa committed:
Major Refactor - Version 0.2.0 (Breaking Changes)

This PR introduces a complete refactor of the Ultimaker integration:

### 🔧 Changes
- UI-based setup with `config_flow`
- Full rewrite using `DataUpdateCoordinator`
- New sensor structure with 60+ entities
- Camera support (stream + snapshot)
- Update entity for firmware
- Entity registry settings improved
- Device grouping, icon updates, translations

### ⚠️ Breaking Changes
- YAML configuration removed
- Old entities may be renamed or removed
- Users must remove the previous integration and re-add it via the UI

Tested on Ultimaker S5 with firmware 9.x.
Co-authored with @alnavasa.

alnavasa committed:
Major Refactor - Version 0.2.0 (Breaking Changes)

This PR introduces a complete refactor of the Ultimaker integration:

### 🔧 Changes
- UI-based setup with `config_flow`
- Full rewrite using `DataUpdateCoordinator`
- New sensor structure with 60+ entities
- Camera support (stream + snapshot)
- Update entity for firmware
- Entity registry settings improved
- Device grouping, icon updates, translations

### ⚠️ Breaking Changes
- YAML configuration removed
- Old entities may be renamed or removed
- Users must remove the previous integration and re-add it via the UI

Tested on Ultimaker S5 with firmware 9.x.
Co-authored with @alnavasa.

alnavasa committed:
Create en.json

Translation requiere for visual interface

alnavasa committed:
Major Refactor - Version 0.2.0 (Breaking Changes)

This PR introduces a complete refactor of the Ultimaker integration:

### 🔧 Changes
- UI-based setup with `config_flow`
- Full rewrite using `DataUpdateCoordinator`
- New sensor structure with 60+ entities
- Camera support (stream + snapshot)
- Update entity for firmware
- Entity registry settings improved
- Device grouping, icon updates, translations

### ⚠️ Breaking Changes
- YAML configuration removed
- Old entities may be renamed or removed
- Users must remove the previous integration and re-add it via the UI

Tested on Ultimaker S5 with firmware 9.x.
Co-authored with @alnavasa.

alnavasa committed:
Update README.md Major Refactor - Version 0.2.0 (Breaking Changes)

This PR introduces a complete refactor of the Ultimaker integration:

### 🔧 Changes
- UI-based setup with `config_flow`
- Full rewrite using `DataUpdateCoordinator`
- New sensor structure with 60+ entities
- Camera support (stream + snapshot)
- Update entity for firmware
- Entity registry settings improved
- Device grouping, icon updates, translations

### ⚠️ Breaking Changes
- YAML configuration removed
- Old entities may be renamed or removed
- Users must remove the previous integration and re-add it via the UI

Tested on Ultimaker S5 with firmware 9.x.
Co-authored with @alnavasa.

alnavasa committed:
Delete custom_components/ultimaker/es.json

located in custom_components/ultimaker/translations/es.json

alnavasa committed:
Delete custom_components/ultimaker/en.json

located in custom_components/ultimaker/translations/en.json

alnavasa committed:
More languages added

More languages added

alnavasa commented:
## Summary

This PR introduces a **major refactor** of the Ultimaker integration for Home Assistant, bumping the version to **0.2.0**.

### 🚀 What's New
- Full migration to **UI-based configuration** (`config_flow`)
- All entities now grouped under a proper **device** with diagnostics
- Added new **camera entity** (stream + snapshot)
- Added **update entity** for firmware version tracking
- Added **diagnostic sensors**: memory usage, uptime, hardware revision, network info, etc.
- Added **multilingual support**: `en`, `es`, `fr`, `de`, `it`, `nl`
- Over **60 new or improved sensors** in total

### ⚠️ Breaking Changes
- YAML-based configuration is no longer supported
- Users must remove old setup and re-add the integration via the UI
- Entities have been renamed and restructured for consistency and maintainability

### 📸 UI Enhancements
A Lovelace example and updated visuals are provided in the new `README.md` and `resources/`.

---

**Original work by [@jellespijker](https://github.com/jellespijker)**  
**Enhanced by [@alnavasa](https://github.com/alnavasa)**

Feedback and testing are welcome!

alnavasa commented:
## Feedback and testing are welcome!

alnavasa commented:
Hi @jellespijker,

First of all, I want to say thank you for your original work on this integration. It’s clear that it was built with care and experience, and I mean no offense at all for having rewritten so much of the code. I simply had some free time and a lot of love for my Ultimaker — it’s an amazing machine.

What started as a personal cleanup for my own setup evolved into something more structured and maintainable, and I figured: if this project is open source, and Ultimaker itself was built on open values from its wooden-frame beginnings, why not contribute back?

This refactor aims to make the integration cleaner, easier to expand, and more aligned with current Home Assistant development standards (UI config flow, device grouping, diagnostics, etc.).

Happy to receive any feedback — and again, full credit to your original work and the spirit of open source that made all of this possible.

Best regards,
[@alnavasa](https://github.com/alnavasa)

jellespijker commented:
Hi @alnavasa no worries, no offense taken, on the contrary, I love that you put the effort in this project.

As luck will have it I have a free day tomorrow. I will put in a bit of effort in reviewing this PR tomorrow.

alnavasa commented:
Thanks so much @jellespijker! I really appreciate your openness and time to review this.
Happy to iterate on anything you suggest — and truly excited to help push this forward together. 

Currently tested on the Ultimaker s5 first hardware revision (0) and working with no errors.

alnavasa commented:
I see a small issue with the version update detection, since one version is detected via text and the other via number, 
Because I really need an update I didn't see this mistake.
I am working on fixing it.

<img width="875" alt="Captura de pantalla 2025-05-29 a las 9 25 28" src="https://github.com/user-attachments/assets/abcd8273-cb1d-4040-aaab-2a3f65a87127" />




Also future enhancement is to implement real translation for all entities, not just the setup menu translations. But I'll leave that for future moths, since it is not critical.

alnavasa committed:
fix: normalize firmware string version check

- Fix mismatch between string and number in firmware version comparison
- Ensure consistency in coordinator data keys
- Improve stability of firmware update entity
- Minor cleanup in sensor.py

alnavasa commented:
###  Additional Fixes

- Resolved inconsistency between `firmware_version` (string) and `latest_firmware` (quoted string)
- Adjusted coordinator to ensure proper structure and stability
- Refined `update.py` entity logic for accurate update_available detection
- Minor improvements in `sensor.py` for better diagnostics handling

Before:
![Captura de pantalla 2025-05-29 a las 9 52 06](https://github.com/user-attachments/assets/a08ae678-b1da-48e1-9b58-85c78ce63efd)

After:
![Captura de pantalla 2025-05-29 a las 9 52 24](https://github.com/user-attachments/assets/782ff38e-5dc1-4233-a0da-3f7a9bba6921)




Thanks again @jellespijker for the base! Let me know if you want anything split into a separate commit.

alnavasa commented:
New logos updated at Home assistant "brands" repository
https://github.com/alnavasa/brands/pull/2

alnavasa committed:
Update README.md

Added new Card example for Lovelace

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
 
-MIN_TIME_BETWEEN_UPDATES = timedelta(seconds=10)
-BASE_URL = "http://{0}/api/v1"
+def get_mac_from_ip(ip):
```
DRY (Don't Repeat Yourself)
Reuse the utility function from camera.py

alnavasa replied:
Solved and added at the coordinator

https://github.com/jellespijker/home-assistant-ultimaker/pull/30/commits/2d7ed51cf6cdc7cd1c48c170e5ba9c33d5dacdd5

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
-    """Representation of a Ultimaker status sensor"""
+_LOGGER = logging.getLogger(__name__)
+DOMAIN = "ultimaker"
+SCAN_INTERVAL = timedelta(seconds=10)
```
Possible future feature request ;-) Make this configurable via the UI. But let's not increase the scope of this PR.

alnavasa replied:
Done!
https://github.com/jellespijker/home-assistant-ultimaker/pull/30/commits/2d7ed51cf6cdc7cd1c48c170e5ba9c33d5dacdd5
+
https://github.com/jellespijker/home-assistant-ultimaker/pull/30/commits/7dd01e1fce4d2331e31429179373de3ab733386d

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+    },
+    "max_temperature_exposed": {
+        "name": "Hotend max temperature",
+        "unit": "°C",
```
```suggestion
        "unit": UnitOfTemperature.CELSIUS,
```

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+
+    "ambient_temperature": {
+        "name": "Ambient Temperature",
+        "unit": "°C",
```
```suggestion
        "unit": UnitOfTemperature.CELSIUS,
```

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+            identifiers={(DOMAIN, entry_id)},
+            connections={("mac", mac_address)} if mac_address else set(),
+            name=user_prefix,
+            manufacturer="Ultimaker",
```
```suggestion
            manufacturer="UltiMaker",
```

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+    },
+    "led_brightness": {
+        "name": "LED brightness",
+        "unit": "%",
```
```suggestion
        "unit": PERCENTAGE,
```

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+    },
+    "led_saturation": {
+        "name": "LED saturation",
+        "unit": "%",
```
```suggestion
        "unit": PERCENTAGE,
```

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+    },
+    "time_elapsed_raw": {
+        "name": "Time elapsed (raw)",
+        "unit": "s",
```
```suggestion
        "unit": UnitOfTime.SECOND,
```

jellespijker commented on custom_components/ultimaker/sensor.py:
```diff
@@ -1,272 +1,580 @@
+
+    "time_remaining_raw": {
+        "name": "Time remaining (raw)",
+        "unit": "s",
```
```suggestion
        "unit": UnitOfTime.SECOND,
```

jellespijker commented on README.md:
```diff
@@ -1,89 +1,225 @@
+[![hacs\_badge](https://img.shields.io/badge/HACS-Custom-orange.svg)](https://hacs.xyz/)
 
-Adds support for the following ultimaker printer sensors:
+Custom integration for monitoring **Ultimaker 3**, **S3**, **S5**, and **S7** 3D printers from Home Assistant. Provides detailed real-time data, camera stream, and firmware updates.
```
Should also work for the S8

alnavasa replied:
Added with https://github.com/jellespijker/home-assistant-ultimaker/pull/30/commits/8f93e1d49b6f24d96f4316f19a37c3f6603c18bc

jellespijker commented on custom_components/ultimaker/camera.py:
```diff
@@ -0,0 +1,81 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+def get_mac_from_ip(ip):
```
Don't assume that every system on hich home assistant runs has these os utils installed.
It might be better to use an existing and dedicated Python library for this. A quick internet search tells me tat getmac https://pypi.org/project/getmac/ is an interresting candidate.

alnavasa replied:
Solved:
https://github.com/jellespijker/home-assistant-ultimaker/pull/30/commits/2d7ed51cf6cdc7cd1c48c170e5ba9c33d5dacdd5

jellespijker commented on custom_components/ultimaker/manifest.json:
```diff
@@ -1,10 +1,16 @@
 {
   "domain": "ultimaker",
   "name": "Ultimaker",
+  "version": "0.2.0",
```
breaking change warrant's a major version bump.

Let's make it version `1.0.0`

alnavasa replied:
Done

jellespijker reviewed (changes requested):
I love it!

You really put in some effort and I and really appreciate the Pull Request. The code quality is pretty decent so I made the assumption, that you have a familiarity with codeing and the review process. If this isn't the case don't hesitate to ask for clarifications.

I didn't test it yet, but if this is going to be a breaking change let's do a major version bump make it 1.0.0 and change the display names of `Ultimaker` to `UltiMaker` as well.

Most of the remarks are small, the one thing that makes me a bit hesitant is the get_mac_from_ip function, that relies on OS tools. Try making that OS agnostic. You can't assume HA is only installed on systems with these tools available. Best use a third-party library for that.

Let me know if you need help.

jellespijker commented:
I can ask some of my colleagues to test it with their Home Assistants: @sergii-shevchenko-ultimaker

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa committed:
Update custom_components/ultimaker/sensor.py

Co-authored-by: Jelle Spijker <spijker.jelle@gmail.com>

alnavasa commented:
Hi @jellespijker, thank you so much for the kind words and your thoughtful feedback!

I’ll definitely take a look into the MAC address utility and try to make it OS-agnostic using a suitable cross-platform library. I totally understand your concern — relying on system tools can be problematic in diverse Home Assistant setups.

As a fun side note: I have no formal coding background — most of this was done with a lot of help from ChatGPT. Apart from writing some YAML for Home Assistant and ESPHome, and building a fully automated Instagram bot that posts sunset photos and timelapses from my home (all running on a Linux container), I’m just someone who enjoys exploring and tinkering with tech.

This PR took around 18 hours in total, filled with trial and error. One of the tricky parts was that ChatGPT insisted there was no way to get the MAC address from within the code directly, hence the OS-dependent call — but now that I better understand the issue, I’ll give it a proper fix.

The MAC address was mainly added so that the Ultimaker integration doesn’t get duplicated in the device registry when also using for example the router integration — both pointing to the same physical device.

Thanks again for the review, and I’m excited to continue polishing suggested changes!

alnavasa committed:
Version changed

Version changed

alnavasa committed:
Update README.md

added S8 and Ultimaker -> UltiMaker

alnavasa committed:
Update info.md

update version numbers

alnavasa committed:
Config flow for seconds interval + MAC address new method

```
### Changes in `config_flow.py`
- Modernized the config flow schema using `vol.Schema` with `selector.TextSelector` and `selector.NumberSelector`.
- Added a new configurable field: `scan_interval`.
- Allows the user to define the update interval in seconds (range: 5–60, default: 10).
- Uses a numeric input with arrows (`NumberSelectorMode.BOX`) for better UX.
- `scan_interval` is now stored in `self.data` and made available to the integration.
- Implemented an options flow to allow updating `scan_interval` from the UI after initial setup.

---

### Make scan interval configurable via UI
- Added `scan_interval` (5–60s, default 10) to `config_flow.py` using `NumberSelector`.
- Passed `scan_interval` from config entry to `UltimakerDataUpdateCoordinator` via `__init__.py`.
- Modified `coordinator.py` to accept and apply the user-defined `scan_interval` using `update_interval=scan_interval`.

---

### Centralize MAC address resolution in coordinator and reuse across all entities
- Added MAC address resolution using `getmac` in coordinator `_async_update_data`.
- Stored result in `coordinator.data["mac"]` for reuse.
- Refactored camera, update, and sensor entities to read MAC from coordinator.
- Replaced all direct calls to `get_mac_from_ip` with coordinator-based access.
- Ensures consistent, efficient, and platform-independent MAC handling.

---

### Add MAC and IP diagnostic sensors, refactor sensor class to support `value_fn`
- Added two new sensors: `ip_address` and `mac_address`
  - Both marked as diagnostic.
  - Disabled and hidden by default.
- Extended sensor definition system to support `value_fn` in addition to `path`.
- Updated `native_value` logic to handle sensors with:
  - `value_fn` (custom value extraction),
  - `path + transform` (existing structure).
- MAC address is now fetched once in coordinator and reused across all entities.
```

alnavasa committed:
Update translation files to include scan_interval

Update translation files to include scan_interval

- Added "scan_interval" label to all existing translation files (en, es, de, fr, it, nl)
- Ensured consistent labeling across all languages
- Keeps config flow translations in sync with new UI fields

alnavasa commented:
@jellespijker  
All changes have been implemented and are working after resolving several errors along the way.  
There's still room for cleanup and making the code more elegant, no doubt.  

Not to aim too high, but maybe we could become the third 3D printer integration to be officially included in Home Assistant?  
Right now, only OctoPrint and Dremel are supported out of the box.

Of course, for that to happen, we'd likely need to add control capabilities, which would probably require proper authentication and credential handling.

alnavasa commented:
Some errors pop un on the log when the machine is powered off, I don't know if this is considered normal and appropriate or a power off state and error correction should be implemented. I recall seeing errors in HA for example if a preconfigured IP camera is powered off

Error 1
```
Este error se originó a partir de una integración personalizada.

Registrador: custom_components.ultimaker.coordinator
Fuente: helpers/update_coordinator.py:412
integración: Ultimaker (documentación, problemas)
Ocurrió por primera vez: 02:43:58 (1 ocurrencia)
Última vez registrado: 02:43:58

Error fetching ultimaker data: Error comunicando con la API de Ultimaker: Connection timeout to host http://192.168.xxx.xxxx/api/v1/printer
```
Error 2
```
Registrador: homeassistant.helpers.event
Fuente: helpers/template.py:646
Ocurrió por primera vez: 02:38:47 (18 ocurrencias)
Última vez registrado: 02:44:36

Error while processing template: Template<template=(**Hora finalización:** {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }} **Tiempo transcurrido:** {{ states('sensor.ums5_time_elapsed') }} h **Tiempo restante:** {{ states('sensor.ums5_time_remaining') }} h **Archivo actual:** {{ states('sensor.ums5_print_job_name') }} **Estado de la impresora:** {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} % **Hotend 1:** {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }} **Hotend 2:** {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }} **Cama caliente:** {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C) **Firmware:** {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}) renders=54>
Error while processing template: Template<template=(**Hora finalización:** {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }} **Tiempo transcurrido:** {{ states('sensor.ums5_time_elapsed') }} h **Tiempo restante:** {{ states('sensor.ums5_time_remaining') }} h **Archivo actual:** {{ states('sensor.ums5_print_job_name') }} **Estado de la impresora:** {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} % **Hotend 1:** {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }} **Hotend 2:** {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }} **Cama caliente:** {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C) **Firmware:** {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}) renders=58>
Error while processing template: Template<template=(**Hora finalización:** {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }} **Tiempo transcurrido:** {{ states('sensor.ums5_time_elapsed') }} h **Tiempo restante:** {{ states('sensor.ums5_time_remaining') }} h **Archivo actual:** {{ states('sensor.ums5_print_job_name') }} **Estado de la impresora:** {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} % **Hotend 1:** {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }} **Hotend 2:** {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }} **Cama caliente:** {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C) **Firmware:** {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}) renders=62>
Error while processing template: Template<template=(**Hora finalización:** {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }} **Tiempo transcurrido:** {{ states('sensor.ums5_time_elapsed') }} h **Tiempo restante:** {{ states('sensor.ums5_time_remaining') }} h **Archivo actual:** {{ states('sensor.ums5_print_job_name') }} **Estado de la impresora:** {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} % **Hotend 1:** {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }} **Hotend 2:** {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }} **Cama caliente:** {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C) **Firmware:** {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}) renders=66>
Error while processing template: Template<template=(**Hora finalización:** {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }} **Tiempo transcurrido:** {{ states('sensor.ums5_time_elapsed') }} h **Tiempo restante:** {{ states('sensor.ums5_time_remaining') }} h **Archivo actual:** {{ states('sensor.ums5_print_job_name') }} **Estado de la impresora:** {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} % **Hotend 1:** {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }} **Hotend 2:** {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }} **Cama caliente:** {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C) **Firmware:** {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}) renders=70>
Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2178, in forgiving_as_timestamp
    return dt_util.as_timestamp(value)
           ~~~~~~~~~~~~~~~~~~~~^^^^^^^
  File "/usr/src/homeassistant/homeassistant/util/dt.py", line 155, in as_timestamp
    raise ValueError("not a valid date/time.")
ValueError: not a valid date/time.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 644, in async_render
    render_result = _render_with_context(self.template, compiled, **kwargs)
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2934, in _render_with_context
    return template.render(**kwargs)
           ~~~~~~~~~~~~~~~^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/jinja2/environment.py", line 1295, in render
    self.environment.handle_exception()
    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^^
  File "/usr/local/lib/python3.13/site-packages/jinja2/environment.py", line 942, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
  File "/usr/local/lib/python3.13/site-packages/jinja2/sandbox.py", line 401, in call
    return __context.call(__obj, *args, **kwargs)
           ~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 2181, in forgiving_as_timestamp
    raise_no_default("as_timestamp", value)
    ~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 1972, in raise_no_default
    raise ValueError(
    ...<2 lines>...
    )
ValueError: Template error: as_timestamp got invalid input 'unknown' when rendering template '**Hora finalización:**   {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }}
**Tiempo transcurrido:**   {{ states('sensor.ums5_time_elapsed') }} h
**Tiempo restante:**   {{ states('sensor.ums5_time_remaining') }} h
**Archivo actual:**   {{ states('sensor.ums5_print_job_name') }}
**Estado de la impresora:**   {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} %
**Hotend 1:**   {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }}
**Hotend 2:**   {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }}
**Cama caliente:**   {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C)  
**Firmware:**   {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}' but no default was specified

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 761, in async_render_to_info
    render_info._result = self.async_render(  # noqa: SLF001
                          ~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^
        variables, strict=strict, log_fn=log_fn, **kwargs
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/src/homeassistant/homeassistant/helpers/template.py", line 646, in async_render
    raise TemplateError(err) from err
homeassistant.exceptions.TemplateError: ValueError: Template error: as_timestamp got invalid input 'unknown' when rendering template '**Hora finalización:**   {{ as_timestamp(states('sensor.ums5_eta')) | timestamp_custom('%H:%M', true) }}
**Tiempo transcurrido:**   {{ states('sensor.ums5_time_elapsed') }} h
**Tiempo restante:**   {{ states('sensor.ums5_time_remaining') }} h
**Archivo actual:**   {{ states('sensor.ums5_print_job_name') }}
**Estado de la impresora:**   {{ states('sensor.ums5_printer_activity') }} // {{ states('sensor.ums5_print_job_progress') }} %
**Hotend 1:**   {{ states('sensor.ums5_hotend_1_temperature') }} °C // {{ states('sensor.ums5_hotend_1_temperature_target') }} °C // {{ states('sensor.ums5_hotend_1_id') }}
**Hotend 2:**   {{ states('sensor.ums5_hotend_2_temperature') }} °C // {{ states('sensor.ums5_hotend_2_temperature_target') }} °C // {{ states('sensor.ums5_hotend_2_id') }}
**Cama caliente:**   {{ states('sensor.ums5_bed_temperature') }} °C (objetivo: {{ states('sensor.ums5_bed_temperature_target') }} °C)  
**Firmware:**   {{ states('sensor.ums5_firmware_version') }} → Última: {{ states('sensor.ums5_latest_firmware_version') }}' but no default was specified
```

Error 3
```
Registrador: frontend.js.modern.202505160
Fuente: components/system_log/__init__.py:331
Ocurrió por primera vez: 02:45:00 (1 ocurrencia)
Última vez registrado: 02:45:00

Uncaught error from Safari 18.5 on macOS 10.15.7 TypeError: e is not an Object. (evaluating '"conditions"in e') conditions.reduce<string[]>( (src/panels/lovelace/common/validate-condition.ts:322:24) reduce ([native code]::) conditions.reduce<string[]>( (src/panels/lovelace/common/validate-condition.ts:323:40) reduce ([native code]::) onChange (src/panels/lovelace/common/validate-condition.ts:336:43) this._clearMediaQueries (src/panels/lovelace/cards/hui-card.ts:239:6) connectedCallback (src/panels/lovelace/cards/hui-card.ts:60:9) src/scoped-custom-element-registry.ts:424:45 insertBefore ([native code]::) t (src/lit-html.ts:1546:43)
```


Error 4
```
Este error se originó a partir de una integración personalizada.

Registrador: custom_components.ultimaker.camera
Fuente: custom_components/ultimaker/camera.py:68
integración: Ultimaker (documentación, problemas)
Ocurrió por primera vez: 02:43:30 (40 ocurrencias)
Última vez registrado: 02:46:32

Error fetching snapshot:
Error fetching snapshot: Cannot connect to host 192.168.xxx.xxxx:80 ssl:default [Connect call failed ('192.168.xxx.xxxx', 80)]
```

Error 5
```
Registrador: homeassistant.components.stream.stream.camera.ums5_camera
Fuente: components/stream/__init__.py:489
integración: Stream (documentación, problemas)
Ocurrió por primera vez: 02:43:54 (7 ocurrencias)
Última vez registrado: 02:47:41

Error from stream worker: Error demuxing stream (EXIT, Immediate exit requested, http://192.168.xxx.xxxx/api/v1/camera/0/stream)
Error from stream worker: Error opening stream (ERRORTYPE_113, Host is unreachable, http://192.168.xxx.xxxx/api/v1/camera/0/stream)

```

Edit:

This level of errors are approved according to Home assistant when the devices go off-line

Though we can improve with a sensor specific to detect if the printer is online or off-line 

Also markdown card gives some errors when sensors are not available so probably is not a good example card. It probably should be removed from the readme.MD.

alnavasa committed:
Update README.md

Added future roadmap, not to make this PR innate, this are some of the ideas I have

alnavasa commented:
ETA sensor not working
Review pending

alnavasa committed:
fix(sensor): correct ETA calculation using datetime.now and timedelta

fix(sensor): correct ETA calculation using datetime.now and timedelta

alnavasa commented:
We fixed the calculation of the `eta` sensor, which was previously failing due to incorrect usage of the `datetime` module (`datetime.datetime.now()` instead of just `datetime.now()`).

However, we have two pending questions:

- **Ambient Temperature**: We're unsure if the value provided is reliable or even relevant, as it's not clearly documented by the Ultimaker API. It seems to return data, but further validation is needed to confirm its usefulness.

- **Hardware Revision**: In our test machine (a very early Ultimaker S5 unit), the `hardware.revision` field returns `0`. We're not sure if this is expected or a limitation of older firmware/hardware. We'd appreciate feedback from other users with newer hardware revisions.

alnavasa committed:
Clean up

Delete unused import of Resource. Minor cleanup and code rearrangement. Tested and confirmed working.

jellespijker commented:
Thx for the changes❤️.
I will blocking some time in my agenda to test this next Friday. GitHub copilot has a couple of decent suggestions you might want to look at

If I don't encounter any issues during testing I will merge this!

alnavasa commented:
Hi @jellespijker 
Great news for Friday, I'll try to fix those issues and test them before updating the code before Friday 

I've already had a look at some of them!

alnavasa commented:
# Copilot Review Notes – Summary for PR

Quick summary of the suggestions Copilot made. No changes applied yet — just capturing ideas for after your review on Friday.

---

## ✅ Reasonable Suggestions (Postpone to 1.0.1?)

These are good, but I don’t have time to test with the machine and implement before Friday:

- Narrow down exception catching in `utils.py` to `(OSError, ValueError)` instead of `Exception`.
- Improve IP validation in `config_flow.py` with regex for IPv4/IPv6.
- Remove `packaging` from `requirements` — not used in the code.
- Translate the Spanish docstring in `coordinator.py` to English for consistency.
- Translate error messages to English (same file) to match HA standards.

All these are valid and would make sense in a follow-up version (`1.0.1`). I didn’t want to touch anything critical or introduce potential regressions this close to review.

---

## ❌ Skipping (For Now or By Choice)

- **IP regex validation** — makes sense, but could introduce breakage if we don’t test thoroughly. Would prefer to handle it with UI selectors or validate upstream.
- **`packaging` removal** — good catch, but we’ll remove in next patch. (after testing)

---

## Thoughts

We can tag this as `1.0.0` or `1.0.0-beta`, and treat these as minor cleanups later.

I didn’t apply any of these changes yet — no time before your code review Friday. Let’s keep it stable for now, polish later, maybe release leter too if you prefer, I am full of work fro 2 weeks now.

---

jellespijker commented:
I didn't manage to get to this. Will look into it this extended weekend (sat, sun, mon).

Apologies for the wait, but we will get there.
